[영화 리뷰] 미션 제출합니다

ImxYJL opened the pull request:
## 리팩토링 할 것들
- 스타일 ㅠㅠ
- 반응형.............
- 에러 메세지 안내하기

~~배포는 리팩토링 완료하고 하겠읍니다 😂~~
[배포 링크](https://imxyjl.github.io/react-movie-review/)

## 1/26 구현 및 1차 리팩토링
### 도입해본 것
#### Jotai 사용 이유
  - 검색어를 Header 쪽에서 받아서 List로 넘겨 줘야 해서 (그치만 drilling은 안 심합니다ㅎㅎ)
  - 겸사겸사 Jotai에서 코드 한 줄로 **웹 스토리지와 연동**할 수 있는 기능을 제공해서 바로 도입했습니다. 
  - 다른 전역 상태 관리 라이브러리들과의 비교
    - Recoil은 Jotai와 비슷하지만 스토리지 연동 기능이 없는 걸로 알고 있고(정확히는 모름), atom Key를 추가해야 하는 등 추가 코드를 작성해야 해서 제외했습니다. 
    - Zustand는 setter를 직접 만들어줘야 해서 제외했습니다. 
    - Jotai의 단점은 상태 관리 로직이 컴포넌트에 분산된다는 것이지만, 작은 규모의 간단한 프로그램이라 사용했습니다. **세터를 별도로 작성할 필요도 없고 파생 상태도 쉽게 만들 수 있어서 타임어택 구현에 좋은 것 같아요.** (단 유지보수 난이도는...)

- infiniteQuery 하나로 두 가지 데이터 가져오기
  - 일반 영화 목록과 검색한 영화 목록의 반환 타입이 같길래 한번 하나의 infiniteQuery로 받아와봤습니다.  

### 에러 메세지
#### 에러 문구
  - 요구사항에 사용자에게 에러 상황을 안내하라? 같은 게 있었던 것 같은데, 사용자에게 있어 유의미한 정보는 
    1. 대략적으로 왜 안 되는지
    2. 뭘 해야 하는지 안내(사실상 새로고침, refetch 정도겠지만)
    3. 재시도/ 홈으로 돌아가기 버튼 등 부가 기능
    일 뿐, 404니 403이니 나누는 것은 일반 사용자에게 큰 의미가 없을 것 같아 투박하게 500/400 에러만 나눴습니다. 로그인처럼 인증이 있는 상황이라면 관련 문구를 더 추가할 수는 있겠네요.

#### 에러 바운더리
  - 처음에는 토스트로 안내할까 했다가 API 요청이 실패한 경우 헤더 외에는 보여줄 데이터가 없어 fallback만 띄웠습니다.
    - 그래서 무의미한 토스트 코드가 있는데 무시해주세요...ㅎㅎ
  - 그런데 지금 생각해보니 무한 스크롤이라 n번째 요청에 실패했을 때는 토스트를 띄워주면 좋겠네요ㅋㅋㅋㅋㅋ

+ 헤더 반응형은 하지 않았읍니다 😶

### 아쉬운 점
#### 스켈레톤
크게 두 가지 아쉬움이 있었어요.

** 1. 유저에게 항상 스켈레톤 보여주기 **
첫 로딩일 때도, 이어서 데이터를 받아올 때도 스켈레톤을 보여주고 싶었는데 isLoading과 isFetchingNextPage를 같이 사용하지 않으면 어느 한 케이스에서는 스켈레톤이 나오지 않더라구요... 그래서 List 코드가 길어졌읍니다.

** 2. 서버에서 받아오는 배열 개수만큼의 스켈레톤 아이템 로딩하기 **
무한 스크롤이라 서버에서 데이터를 실제로 받아오기 전까지는 새 페이지에 몇 개의 데이터를 받았는지 알 수 없어서, 기본 사이즈인 20개를 렌더하고 있습니다.
딱 맞는 개수로 보여주겠다고 데이터를 받고 나서야 스켈레톤을 보여주는 건 의미가 없다고 생각해서요. 혹시 이 문제를 해결하셨는지 궁금합니다!!
 
### 후기
- 여태까지 props 넘겨주는 방식만 사용했는데 상태 관리 라이브러리를 쓰니까 확실히 편하다...
- 생각보다 기능 구현은 쉬웠는데(옵저버 훅은 가져왔지만...^^) CSS에서 애를 먹었다.
  - 특히 스켈레톤에서 시간을 많이 뺏겼다. 
  - 다른 것들은 크게 어렵지 않았지만, 이미 있는 css에 뭔가를 추가하려고 하니까 디버깅하는 시간이 필요했다.

ImxYJL committed:
chore: 초기 환경설정

ImxYJL committed:
feat: Header, Search 컴포넌트

ImxYJL committed:
chore: jpg 타입 추가

ImxYJL committed:
chore: global 스타일 조정

ImxYJL committed:
chore: assets 세팅

ImxYJL committed:
feat: 영화 목록 타이핑

ImxYJL committed:
feat: 영화 목록을 불러오는 api 함수 및 훅

ImxYJL committed:
feat: 무한 스크롤을 사용하기 위한 훅

ImxYJL committed:
chore: GlobalStyle 적용

ImxYJL committed:
feat: MovieList 컴포넌트

ImxYJL committed:
feat: MovieItem 컴포넌트

ImxYJL committed:
chore: Jotai 다운로드

ImxYJL committed:
feat: 영화 목록 불러오는 함수 리팩토링 및 영화 검색 API 호출 함수 추가

ImxYJL committed:
feat: 검색어를 저장하는 atom 및 검색어 설정 기능 추가

ImxYJL committed:
feat: 검색어 유무에 따라 영화 목록 데이터를 불러오는 쿼리 훅

ImxYJL committed:
feat: 엔드포인트 상수 변경

ImxYJL committed:
refactor: 쿼리 재호출 횟수 1로 제한

ImxYJL committed:
feat: 영화 목록 상태에 따른 Title 문구 변경

ImxYJL committed:
feat: 모달을 위한 공통 컴포넌트 및 훅

ImxYJL committed:
feat: 모달 훅 추가

ImxYJL committed:
feat: 영화 상세 정보를 불러오는 함수 및 쿼리 훅

ImxYJL committed:
feat: StarButton 컴포넌트

ImxYJL committed:
feat: 별점을 줄 수 있는 컴포넌트

ImxYJL committed:
feat: 상세 정보 모달

ImxYJL committed:
feat: App에 컴포넌트 렌더링

ImxYJL committed:
feat: 나의 별점을 로컬 스토리지에 연동하는 기능 추가 및 별점 관련 기능을 훅으로 분리

ImxYJL committed:
feat: 영화 목록 가운데 정렬

ImxYJL committed:
feat: 스켈레톤 추가

ImxYJL committed:
fix: 모달 클릭 시 새로고침 방지

ImxYJL committed:
refactor: 모달 스타일 조정

ImxYJL committed:
feat: 영화 목록 및 모달 반응형

ImxYJL committed:
feat: 로고를 클릭하면 새로고침되는 기능

ImxYJL committed:
fix: isLoading과 isFetchingNextPage 분리

ImxYJL committed:
refactor: Header 태그 구조 변경

ImxYJL committed:
feat: 누락됐던 타입 업로드...

ImxYJL committed:
refactor: 자잘한 스타일 수정들

ImxYJL committed:
feat: 에러 바운더리, 에러 fallback 추가

ImxYJL committed:
feat: 검색 결과 안내

ImxYJL committed:
refactor: suspenseQuery 제거

ImxYJL committed:
refactor: 영화 상세 정보 모달 관련 상태 훅으로 분리

ImxYJL committed:
chore: 사용하지 않는 매개변수 삭제

chosim-dvlpr commented on src/App.tsx:
```diff
@@ -1,5 +1,29 @@
+    <div style={AppStyle}>
+      <Header />
+
+      <QueryErrorResetBoundary>
+        {({reset}) => (
+          <ErrorBoundary fallback={ErrorFallback} resetQueryError={reset}>
+            <main>
+              <MovieList />
+            </main>
+          </ErrorBoundary>
+        )}
+      </QueryErrorResetBoundary>
```
와우 에러바운더리까지...!👍🏻

ImxYJL replied:
사실 있는 코드를 가져왔습니다ㅋㅋㅋㅋ

chosim-dvlpr commented on src/hooks/useGetMovieDetail.ts:
```diff
@@ -0,0 +1,14 @@
+import {getMovieDetail} from '../apis/movie';
+
+const useGetMovieDetail = (movieId: number) => {
+  const result = useQuery({
+    queryKey: ['movieDetail', movieId],
+    queryFn: () => getMovieDetail(movieId),
+    staleTime: 24 * 60 * 60 * 1000, // 24시간
+  });
```
여기서만 따로 staleTime을 지정한 이유가 있을까요?

ImxYJL replied:
실시간 인기 영화 순위는 시간별로 순위 변동이 있을 수 있지만 상세 정보는 staleTime을 설정해도 될 것 같다는 생각이 들었어요.
평점이 조금 걸리는데 평점은 단시간에 변할 것 같지 않아서 넉넉하게 하루로 설정해봤어요!

chosim-dvlpr commented on src/components/DetailModal/index.tsx:
```diff
@@ -0,0 +1,57 @@
+}
+
+const DetailModal = ({selectedMovieId, closeModal}: DetailModalProps) => {
+  const {data} = useGetMovieDetail(selectedMovieId);
+  if (!data) return null;
```
모달 데이터를 불러오고 있는 중에는 모달이 안띄워지더라구요. 그래서 Movie를 클릭했을 때 약간의 지연시간 후에 모달이 띄워지고 있구요! 이 부분에서 사용성을 위해 스켈레톤이나 Suspense같은 UI를 넣는 것은 어떨까요?

ImxYJL replied:
사실 저는 1초 언저리의 짧은 로딩 시간이라면 스켈레톤을 거는 게 오히려 정신없다고 생각해서 선호하지 않는 편인데, 모달은 1초대의 로딩이지만 화면 전환이 바로 일어날 것을 기대하고 클릭하는 거니까 모달은 바로 띄우되 로딩 UI를 보여주는 게 훨씬 자연스러울 것 같네요! 
 
개발하면서 눌렀던 것만 계속 눌러서 느린 줄도 모르고 있었네요...ㅋㅋㅋㅋㅋ

chosim-dvlpr commented on src/utils/error.ts:
```diff
@@ -0,0 +1,4 @@
+export const createApiErrorMessage = (statusCode: number) => {
+  if (statusCode >= 500) return '서버 에러가 발생했어요. 잠시 후 다시 시도해주세요.';
+  if (statusCode >= 400) return '잘못된 요청입니다.';
+};
```
오👍🏻👍🏻👍🏻

ImxYJL replied:
투박하지만 😅 한창 이 미션 할 때는 케이스별로 메세지를 다르게 안내했지만... 일반 사용자에게는 그래서 어쩌라고? 일 것 같아 그냥 심플하게 나눴습니다ㅎㅎ

chosim-dvlpr commented on src/apis/movie.ts:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  const data = await response.json();
+  return data as MovieDetail;
```
여기서는 as로 리턴하고 있군요! 혹시 타입 단언을 하는 경우도 `getMovieDetail` 함수의 리턴값 타입 추론이 되나요??(잘 몰라서...) 사실 서버로부터 오는 데이터 타입을 Typescript가 추론할 수는 없어서 `as`도 상관없을 것 같지만 권장되는 형식은 명시적인 반환 타입 선언을 하는 방법 같아서요! 올리는 as와 반환 타입 선언 중 어떤 방법이 더 낫다고 생각하는지 궁금해요!

ImxYJL replied:
이 부분은 프로젝트 할 때 계속 써와서 그런가 별 생각 없이 썼어요...ㅎ (큰 문제가 있었다면 누군가 지적해줬겠지 하는 안일함) 당장 드는 생각은 일반적인 경우에는 단언을 해도 문제는 없을 것 같다는 생각입니다. 하지만 서버가 깨져서 이상한 값을 보내주는 경우에는 디버깅하기 어려울 수도 있을 것 같아요. 

타입 추론은 as가 애초에 이걸로 추론해라! 라고 명령하는 거니까 문제없이 되는 것 같습니다. 실제로 리액트 쿼리 코드를 쓸 때 제네릭으로 리턴 타입을 명시해주지 않아도 알아서 data 타입을 추론해주더라고요.  

만약 고쳐본다면 Promise에 제네릭으로 타입을 걸어줄 것 같습니다~!

chosim-dvlpr commented on src/hooks/useStarRating.ts:
```diff
@@ -0,0 +1,69 @@
+  const [starButtonStateList, setStarButtonStateList] = useState<StarButtonState[]>(INITIAL_STAR_BUTTON_LIST);
+  const [selectedIndex, setSelectedIndex] = useState(0);
+
+  // 초기 로드 시 로컬 스토리지에서 별점 값 불러오기
+  useEffect(() => {
+    if (myRateList) {
+      const savedRate = myRateList.find(rate => rate.movieId === movieId);
+      if (savedRate) {
+        setSelectedIndex(savedRate.starButtonIndex); // 저장된 별점 인덱스 설정
+      }
+    }
+  }, [myRateList, movieId]);
```
useEffect를 쓰지 않고 selectedIndex를 선언할 때 바로 이 로직을 넣어줄 수도 있을 것 같은데.. 이 방법은 어떨까요?

ImxYJL replied:
앗 제가 주석을 이상하게 적어놨는데 😶😶 정확히는 초기 로드에서만 사용하는 로직이 아니라서 useEffect가 필요하긴 해요.

처음에 로컬 스토리지와 완전히 연동하기 전에는 myRateList가 null일 수 있고, movieId가 변할 때마다 selectedIndex를 바꿔줘야 하기 때문입니다! 

주석 수정하고, `if(myRateList)`문 삭제하는 대신 myRateList?. 스타일로 리팩토링할게요!

chosim-dvlpr commented on src/hooks/useStarRating.ts:
```diff
@@ -0,0 +1,69 @@
+  const handleStarButtonClick = (index: number) => {
+    setSelectedIndex(index);
+
+    // 로컬 스토리지 업데이트
+    setMyRateList(prev => {
+      const updatedList = prev
+        ? [...prev.filter(rate => rate.movieId !== movieId), {movieId, starButtonIndex: index}]
+        : [{movieId, starButtonIndex: index}];
+
+      return updatedList;
+    });
```
오홍 로컬 스토리지를 사용했군요! 이번 구현에서 이 방법을 사용한 이유가 있을까요??

ImxYJL replied:
별도로 API를 구현할 생각이 없었고, 또 별도로 DB나 자체 API를 구축하지 않는 이상 유저가 매긴 평점은 실제로 내려받는 평점 데이터에 영향을 주지 않는 개인적인 데이터라 그냥 로컬에 저장했습니당

chosim-dvlpr commented on src/components/DetailModal/index.tsx:
```diff
@@ -0,0 +1,57 @@
+              <S.Overview>{overview}</S.Overview>
+
+              <S.MyRatingWrapper>
+                <MyRating movieId={selectedMovieId} />
```
selectedMovieId도 props가 꽤 깊이 있어서 전역으로 관리해도 좋을 것 같은데 어떨까요??

ImxYJL replied:
지금 뎁스가 3이고, 각 계층에서 id를 단독으로 사용하는 로직이 있어서(단순 props 전달만 하는 컴포넌트 없음) 그냥 props로 넘겨받고 있습니다!

chosim-dvlpr commented on src/hooks/useGetList.ts:
```diff
@@ -0,0 +1,22 @@
+
+  const result = useInfiniteQuery({
+    queryKey: searchText ? ['searchedMovieList', searchText] : ['movieList'],
+    queryFn: ({pageParam = 1}) =>
+      searchText ? getSearchedMovieList({pageParam, title: searchText}) : getMovieList({pageParam}),
```
queryFn 호출 부분이 삼항 연산자로 되어있는데 이 부분을 따로 분리해보면 어떨까요?

ImxYJL replied:
FE 개발에서는 삼항 연산자가 두루 쓰이는 것 같아서, 저는 알아보기 힘들어지거나 eslint에서 고치라고 하는 코드가 아니라면 삼항 연산자를 애용하긴 해요. (백엔드에서는 거의 금기시된 것 같더라고요?) 이 부분의 가독성이 떨어져서 코멘트 남겨주신 걸까요?!

혹시 대세론이 있다면 알려주세요 🤗

chosim-dvlpr commented on src/components/MovieList/index.tsx:
```diff
@@ -0,0 +1,98 @@
+  const searchText = useAtomValue(searchTextAtom);
+
+  const {data, hasNextPage, fetchNextPage, isLoading, isFetchingNextPage} = useGetList();
+  const movies = data ? data.pages.flatMap(page => page.results) : [];
```
데이터 형식 처리를 MovieList에서 하고 있군요! 제 생각에는 역할에 따라 useGetList에서 담당할 수도 있을 것 같아요!

ImxYJL replied:
마침 프룬 코드에서 평탄화를 훅에서 해 주는 게 좋아 보여서 코멘트 남겼어유ㅎㅎ 바로 도입할게요! 

이 가공 작업의 책임이 어디에 있냐의 문제인 것 같은데, 엄밀히 말하면 훅에서는 리액트 쿼리를 사용해서 순수하게 서버 데이터만 가져오게 하고 가공은 사용처(컴포넌트)에서 알아서 해라~ 식으로 data만 딱 던져주는 게 정석이라고 생각했어요. 

하지만 지금은 사용하는 데이터 형식이 한정되어 있고 페이지네이션이라는 특수 로직이 들어가 있어서 프룬이 제안해준 스타일이 여러모로 좋은 것 같아요! 어차피 컴포넌트는 이런저런 로직으로 정신없으니까요ㅋㅋㅋㅋ

chosim-dvlpr commented on src/components/MovieList/index.tsx:
```diff
@@ -0,0 +1,98 @@
+  const {data, hasNextPage, fetchNextPage, isLoading, isFetchingNextPage} = useGetList();
+  const movies = data ? data.pages.flatMap(page => page.results) : [];
+
+  const lastElementRef = useInfiniteScroll({
+    fetchNextPage,
+    isLoading,
+    isLastPage: !hasNextPage,
+  });
```
오 이렇게 무한 스크롤 로직을 분리하니 정말 깔끔하네요👍🏻

chosim-dvlpr commented on src/hooks/useInfiniteScroll.ts:
```diff
@@ -0,0 +1,31 @@
+  const lastElementRef = useCallback(
+    (node: HTMLElement | null) => {
+      if (isLoading) return;
+      if (observer.current) observer.current.disconnect();
```
메모리 누수 방지를 위해 언마운트시 disconnect를 하면 좋을 것 같은데 올리는 어떻게 생각하실까요??

ImxYJL replied:
필수라고 생각합니다 🫠 추가하러 갈게요!!

chosim-dvlpr commented on src/components/MovieList/index.tsx:
```diff
@@ -0,0 +1,98 @@
+  const handleItemClick = (event: React.MouseEvent<HTMLUListElement>) => {
+    event.preventDefault();
+
+    const listItem = (event.target as HTMLElement).closest('li');
```
closest를 사용한 이유가 있을까요?!

ImxYJL replied:
이벤트 위임을 사용중이라 클릭 이벤트가 일어났을 때 클릭 지점과 가장 가까운 li 요소를 찾아주려고 했습니당

chosim-dvlpr commented on src/components/MovieList/index.tsx:
```diff
@@ -0,0 +1,98 @@
+
+              return (
+                <>
+                  <div
+                    key={movie.id}
+                    ref={isLastMovie ? lastElementRef : null} // 마지막 요소에 ref 연결
+                  >
```
오호 이렇게 마지막 요소에 ref를 넣는 방법도 있군요..! 이렇게 하면 확실히 마지막 요소까지 도달했을 때 새로운 API를 호출할 수 있을 것 같아요

chosim-dvlpr commented on src/components/MovieList/index.tsx:
```diff
@@ -0,0 +1,98 @@
+            movies.map((movie, index) => {
+              const isLastMovie = index === movies.length - 1;
+
+              return (
+                <>
+                  <div
+                    key={movie.id}
+                    ref={isLastMovie ? lastElementRef : null} // 마지막 요소에 ref 연결
+                  >
+                    <MovieItem
+                      key={movie.id}
+                      id={movie.id}
+                      title={movie.title}
+                      poster_path={movie.poster_path}
+                      vote_average={movie.vote_average}
+                    />
+                  </div>
+                </>
```
이 부분을 새로운 컴포넌트로 분리해보는 건 어떨까요? ref를 MovieItem에 직접 넘겨주는 방법으로도 해볼 수 있을 것 같아요

ImxYJL replied:
#### 컴포넌트 분리
무한 스크롤과 관련된 로직까지가 MovieList의 책임이라고 생각했고, div 코드가 3줄짜리라 분리하지 않았습니다!

#### MovieItem에 ref 전달하기
위와 비슷한 맥락으로 MovieItem은 자신이 렌더링할 데이터만 넘겨받아야 한다고 생각해서 혹 컴포넌트를 분리하게 되더라도 ref를 Item에 줄 것 같지는 않아요.

다른 의견이 있다면 알려주세요!

chosim-dvlpr commented on src/components/MovieItem/index.tsx:
```diff
@@ -0,0 +1,24 @@
+import * as S from './styles';
+import {toOneDecimalPlace} from '../../utils/number';
+
+export type MovieItemProps = Pick<Movie, 'id' | 'title' | 'poster_path' | 'vote_average'>;
```
오 Pick을 사용했군요! 올리는 Pick과 Omit의 사용 기준이 있을까요? 저는 Omit을 주로 사용하는 편이라서요!

그리고 MovieItemProps는 다른 곳에서 사용되지 않기 때문에 export가 없어도 될 것 같슴다!

ImxYJL replied:
Movie 인터페이스 타입에 속성이 많은 반면 사용할 속성은 3개뿐이라 Pick을 썼어요. 가져오는 속성이 많거나, 반대로 원본의 속성이 별로 없었으면 Omit을 썼을 거라 어떤 것을 쓸지는 상황에 따라서 달라질 것 같네요! Pick과 Omit의 뉘앙스 차이도 있지만 가장 기본적인 선택 기준은 개수인 것 같아요.

export 부분은 한떄 가볍게 고민해봤었는데요, 개발하면서 외부에서 사용하지 않는 것을 export했을 때 생기는 문제보다 기존에 export가 없던 것들을 새롭게 export한 경험(=귀찮음)이 훨씬 많아서 엄격하게 신경쓰고 있지 않습니다 😂
혹시 성능 이슈같은 게 있을까요?

chosim-dvlpr commented on src/components/MovieItem/index.tsx:
```diff
@@ -0,0 +1,24 @@
+export type MovieItemProps = Pick<Movie, 'id' | 'title' | 'poster_path' | 'vote_average'>;
+
+const MovieItem = ({id, title, poster_path, vote_average}: MovieItemProps) => (
+  <S.MovieItem id={String(id)}>
```
id는 li태그의 id인 걸까요? id태그를 넣어준 이유가 궁금해요!

ImxYJL replied:
movieId입니다! 무한 스크롤 로직이라 이벤트 핸들러를 각 Item에 전부 달아주면 메모리 이슈가 있을 것 같아 위임으로 구현했어요. 부모 컴포넌트인 List에서 클릭 이벤트가 일어나면 가까운 li 요소를 찾아 id를 가져와 구분합니다.

chosim-dvlpr commented on src/components/MovieList/index.tsx:
```diff
@@ -0,0 +1,98 @@
+            })}
+
+          {isFetchingNextPage && <SkeletonMovieList />}
+          {!isFetchingNextPage &&
+            movies.map((movie, index) => {
+              const isLastMovie = index === movies.length - 1;
+
+              return (
+                <>
+                  <div
+                    key={movie.id}
+                    ref={isLastMovie ? lastElementRef : null} // 마지막 요소에 ref 연결
+                  >
+                    <MovieItem
+                      key={movie.id}
+                      id={movie.id}
+                      title={movie.title}
+                      poster_path={movie.poster_path}
+                      vote_average={movie.vote_average}
+                    />
+                  </div>
+                </>
+              );
+            })}
+        </S.MovieList>
```
infiniteQuery에서 새로운 배열을 기존 배열에 넣어주고 있기 때문에 이 부분은 없어도 될 것 같아요~!

ImxYJL replied:
앗 이 부분때문에 영화가 2번 보였던 거네요😂 덕분에 알아차렸어요!

chosim-dvlpr commented on src/components/Search/styles.ts:
```diff
@@ -0,0 +1,34 @@
+import styled from '@emotion/styled';
+import searchImg from '../../assets/search_button.png';
+
+export const Search = styled.section`
```
form으로도 처리해볼 수 있을 것 같은데 올리는 어떤 방법이 더 편할까요?

ImxYJL replied:
이런 검색창은 충분히 약식 form이라고 볼 수 있을 것 같아요. 사용자 입력을 받고 서버에 제출하니까요~!
다만 개인적인 구분으로는 Post나 Patch 요청이 들어가야 진짜(?) form이라고 생각해서 section으로 만들었어요ㅎㅎ

chosim-dvlpr commented on src/components/common/ModalBackground/index.tsx:
```diff
@@ -0,0 +1,16 @@
+interface ModalBackgroundProps {
+  closeModal: (() => void) | null;
+}
+const ModalBackground: React.FC<PropsWithChildren<ModalBackgroundProps>> = ({children, closeModal}) => {
```
React.FC를 사용한 이유가 있을까요?!

ImxYJL replied:
사실 이건 직접 구현한 게 아니라 잘 모르겠네요... 🤣 그때 컨벤션으로나 이 코드의 통일성으로나 FC는 사용하지 않는 게 맞습니다. 섬세프룬~~~

chosim-dvlpr commented on src/apis/movie.ts:
```diff
@@ -0,0 +1,86 @@
+  };
+  const endpoint = `${POPULAR_MOVIE_URL}?${new URLSearchParams(params).toString()}`;
+
+  await delay(1500); // NOTE: 스켈레톤 디스플레이를 위함
```
서버에서 데이터를 호출하는 지연 시간이 있기 때문에 delay를 걸지 않아도 좋을 것 같아요! 저는 구현할 때 네트워크 속도나 performance에서 CPU를 제어하면서 테스트했답니다😊

ImxYJL replied:
저도 throttle을 걸어봤었는데 정말 정직하게 화면만 늦게 뜨고 스켈레톤은 순식간에 지나갔어서 의도적으로 딜레이를 줬었어요. 그런데 지금은 인위적인 딜레이 없어도 스켈레톤이 보여서 제거했어요!

chosim-dvlpr commented on src/components/common/ErrorBoundary/index.tsx:
```diff
@@ -0,0 +1,59 @@
+    const {hasError, error} = this.state;
+    const {children, fallback: FallbackComponent} = this.props;
+
+    // 에러 메세지에 TOAST_ERROR 포함하면 fallback 대상에서 제외
+    const isHandleError = !error?.message.includes(TOAST_ERROR);
```
오홍 토스트 구현도 궁금해지네요🤩 기대해봐도....좋을까요....?ㅋㅋㅋㅋㅋㅋ

chosim-dvlpr commented on src/hooks/useModalClose.ts:
```diff
@@ -0,0 +1,59 @@
+}
+
+const useModalClose = ({closeModal, modalBackgroundRef}: UseModalCloseProps) => {
+  const isNodeElement = (element: EventTarget | null): element is Node => {
+    return element instanceof Node;
+  };
```
와우 반환 타입으로 이렇게 하는 방법도 있군요...!

ImxYJL replied:
정작 저도 이걸 반년만에 봐서 낯서네요ㅋㅋㅋㅋㅋㅋ

chosim-dvlpr commented on src/styles/globalStyles.ts:
```diff
@@ -0,0 +1,31 @@
+  ${reset()}
+
+  * {
+    box-sizing: border-box !important;
```
!important를 사용한 이유가 궁금합니다!

ImxYJL replied:
요것도 그냥 복붙인데 아마 border-box가 만에 하나 적용되지 않으면 가장 기본적인 css 리셋이 안 먹히는 거라 important가 붙은 게 아닐까 싶어요!

chosim-dvlpr reviewed (changes requested):
안녕하세요 올리~~ 구현하느라 고생많으셨습니다!!

> ** 2. 서버에서 받아오는 배열 개수만큼의 스켈레톤 아이템 로딩하기 **
무한 스크롤이라 서버에서 데이터를 실제로 받아오기 전까지는 새 페이지에 몇 개의 데이터를 받았는지 알 수 없어서, 기본 사이즈인 20개를 렌더하고 있습니다.
딱 맞는 개수로 보여주겠다고 데이터를 받고 나서야 스켈레톤을 보여주는 건 의미가 없다고 생각해서요. 혹시 이 문제를 해결하셨는지 궁금합니다!!

저도 스켈레톤 개수를 임의로 지정했어요 ㅋㅋㅋ 다시 생각해봤을 때 볼러올 데이터가 마지막 페이지라든지 여러 이유로 인해 20개 미만인 경우에는 스켈레톤과 데이터의 개수가 일치하지 않겠군요.. 
그렇지만 스켈레톤의 역할을 생각해보자면 사용자에게 `데이터를 불러오고 있어!`라는 메시지를 던지기 위함이라고 생각하는데요, 그래서 저는 스켈레톤 개수와 데이터의 개수가 무조건 일치해야 한다고 생각하지는 않아요! 올리는 어떻게 생각하실지 궁금합니당

배포 사이트를 확인하던 중 아래와 같은 버그를 발견했어요! 다음 리팩터링 때 버그를 고쳐주시면 감사하겠습니다!~
1. MovieList 데이터 호출은 1번만 일어나지만 데이터는 두 번 들어가는 현상
2. 스켈레톤은 보여주고 있지만 API 호출이 뒤늦게 일어나고 있는 문제 (delay가 걸려있지만 delay를 제거하고 구현하면 더 좋을 것 같아요!)

추가로 있으면 좋을 법한 기능들인데요, 검색 시 API를 호출할 때 스켈레톤을 보여주는 기능과 헤더의 반응형도 구현하면 좋을 것 같습니다~!!
(`Search`의 배경색이 검정색인데 흰색으로 변경되면 반응형을 더 잘 확인할 수 있을 것 같아요😉)

고생하셨어요~👏🏻👏🏻

ImxYJL committed:
fix: 중복 영화 목록 삭제

ImxYJL committed:
refactor: 모호한 주석 수정 및 간단한 리팩토링

ImxYJL committed:
fix: observer에 컴포넌트 언마운트시의 disconnect 코드 추가

ImxYJL committed:
chore: 불필요한 React.FC 제거

ImxYJL committed:
refactor: API 요청 시 인위적인 delay 코드 제거

ImxYJL committed:
fix: 영화 목록 렌더링 방식 및 스타일 리팩토링

1. 누락됐던 Thumbnail 컴포넌트 적용
2. movie.length === 0 체크를 상단에서 수행해 영화 검색 시에 스켈레톤이 뜨지 않던 문제 해결
3. 스타일 미세 조정

ImxYJL committed:
refactor: 평탄화를 훅에서 수행

ImxYJL committed:
refactor: 기존 영화 목록 컴포넌트를 MovieListContainer와 MovieList로 분리

ImxYJL committed:
chore: 누락된 styles 파일 업로드

ImxYJL committed:
feat: 모달 스켈레톤 추가

ImxYJL committed:
refactor: 반응형 스타일 수정

ImxYJL committed:
feat: 모달 스켈레톤 코드 추가

ImxYJL committed:
feat: 반응형 헤더

ImxYJL commented:
버그 수정 및 반응형 헤더, 모달 스켈레톤 추가했습니다!
그 외에는 간단한 구조 리팩토링을 진행했어요. 
토스트는 현재 구현상 추가할 곳이 애매해서 일단 넣지는 않았어요.

고정 크기를 줘서 데이터 로드 시의 layout shift는 없앴는데, 이미지가 완전히 표시될 준비가 되기 전까지는 계속 스켈레톤을 보여주고 싶은데 고민이네요... 
[배포 링크](https://imxyjl.github.io/react-movie-review/)

ImxYJL committed:
refactor: QueryErrorBoundary를 별개의 컴포넌트로 분리
